### PR TITLE
[Feature][MRV2] Support lmhead TP in model runner v2 sample path

### PR DESCRIPTION
**Pull request by xuchi-0808**
## Summary

This PR adds model-runner-V2 support for the `lmhead_tensor_parallel_size` knob of fine-grained TP, which previously only worked on the V1 runner.

lmhead TP shards the LM head's vocab dimension across a group of ranks built along the **DP axis** (`vllm_ascend/distributed/parallel_state.py`). The LM-head operator runs `all_gather` → vocab-sharded GEMM → `all_to_all` across that group on **every** `compute_logits` call (`vllm_ascend/ops/vocab_parallel_embedding.py`). For these collectives to terminate, every rank of the group must call `compute_logits` with the same number of rows on every engine step — otherwise the collectives hang (no error).

The V1 runner maintained this invariant on two paths; V2 had neither:

1. **Real sampling** — V1 pads `logits_indices` to a fixed capacity and trims the logits rows before sampling. V2 cannot pad the indices: its sampler receives the whole `input_batch` and gathers penalties inputs by `input_batch.logits_indices` on the fly, so V1-style index padding would silently corrupt penalties.
2. **Idle DP ranks** — V1's dummy run executes `dummy_compute_logits` (`need_dummy_logits`), so ranks without requests still join the LM-head collectives each step. V2's dummy batch only runs the forward; idle ranks never call `compute_logits`, and the whole engine deadlocks.

This PR ports both, adapted to the V2 sampling contract.

## Changes

Two overrides in `vllm_ascend/worker/v2/model_runner.py` (both pure passthroughs to `super()` when the feature is off — zero behavior change for existing deployments):

**1. `sample()` — real sampling path.** Sampled hidden states are gathered by the real `logits_indices`, padded up to the group-agreed capacity (`max_num_reqs * decode_query_len`, derived purely from global config and equal to upstream's own logits capacity bound in `StructuredOutputsWorker`), `compute_logits` runs on the padded rows, then the padding is trimmed so padded rows never reach the sampler or grammar bitmask. A fail-fast assert turns a would-be collective shape desync into a clear error if the capacity formula ever stops matching upstream logits production. The dispatch tail mirrors `GPUModelRunner.sample`; a **dispatch-tail canary unit test** runs upstream `sample()` and the override on the same mocks and asserts identical dispatch call sequences, so main2main drift fails loudly.

**2. `_dummy_run()` — idle DP ranks.** After the upstream dummy forward, `compute_logits` is called once with zero-indexed rows at the same capacity `sample()` pads to (both derive from `_lmhead_tp_max_num_logits()`, so the formula cannot drift), mirroring V1's `need_dummy_logits`. Profiling runs (the profile dummy sampler already calls `compute_logits` on every rank) and non-last-PP ranks (no logits) are excluded.

`tests/ut/worker/test_model_runner_v2_finegrained_tp.py` (new): 9 test items, one per risk path — passthrough, pad-then-trim (parametrized below/at capacity), the capacity fail-fast assert, a 3-combination dispatch-tail canary, and the dummy-run lockstep regression with its negative paths. The canary runs upstream `sample()` and the override on the same mocks and asserts identical dispatch call sequences, which also locks the grammar / rejection-sampler contracts — no separate per-branch tests are needed. Components are autospecced against the real upstream classes so argument drift fails loudly. The whole file runs in ~0.2s and passes with no NPU devices visible (pure CPU mocks).

## Correctness: the deadlock this fixes

A3 16-chip, Qwen3-30B-A3B-W8A8 (`--quantization ascend`), V2, 1P1D (P=TP8 kv_producer / D=DP8 TP1 kv_consumer, MooncakeConnectorV1, `recompute_scheduler_enable=true` — the weekly shape):

| # | Mode | lmhead | Result |
|---|---|---|---|
| 1 | eager | off | ✅ 3/3 correct (control) |
| 2 | eager | 8 | ❌ hung on first request (pre-fix) |
| 3 | eager | 8 | ✅ with this PR — byte-identical to #1 |
| 4 | graph (FULL_DECODE_ONLY) | 8 | ✅ with this PR — 3/3 correct |

Rows 1 vs 2 isolate lmhead TP as the only variable. `py-spy` on the hang: busy rank stuck in the LM-head `all_gather` inside `sample()` → its NPU stream never completes → it drops out of the DP-gloo `all_reduce` (`sync_cudagraph_and_dp_padding`) → idle workers wait on gloo forever. Row 3's byte-identical outputs across repeated rounds (e.g. 17×23 → 391) confirm the sharding/trim math is exact; row 4 confirms the same under ACLGraph.

## Performance: V2 vs V1 parity (same 1P1D shape)

V1/V2 toggled only via `VLLM_USE_V2_MODEL_RUNNER` (both sides rebooted, all other args identical), 128 requests / concurrency 64 / ~2k in / 512 out / temp=0:

| Mode | Metric | V2 | V1 |
|---|---|---|---|
| graph, steady (run 2) | output tok/s | 1600.1 | 1588.8 |
| | TPOT ms | 37.8 | 38.1 |
| graph, incl. cold start (run 1) | output tok/s | 1453.6 | 1444.4 |
| eager | output tok/s | 164.7 | 165.3 |

Differences are <1%, within run-to-run variance; graph capture time matches (2:28 vs 2:27). Expected: the cross-DP LM-head collectives are the same operator code under both runners; the runner difference is only where pad/trim is scheduled.

## Known limitations

- `prompt_logprobs` requests are not supported with lmhead TP (the prompt-logprobs path calls `compute_logits` with per-rank uneven chunk sizes); same limitation as V1. No startup flag can reject the combo — `prompt_logprobs` is a per-request sampling parameter.
- Spec decoding: the sample-path rejection branch is wired and unit-tested, but e2e V2 spec decode + lmhead TP (incl. the draft-side alignment V1 has) is not yet hardware-validated; a separate follow-up PR covers the draft side.
- Pre-existing V1 issue, unchanged by this PR: V1's zero-padded `logits_indices` can corrupt `kv_sharing_fast_prefill` metadata (padded zeros are attributed to request 0; pure-decode D nodes don't hit it). The V2 path never pads `logits_indices`, so this class cannot occur here. A separate V1 mutual-exclusion follow-up is planned.

## Validation

- [x] Unit tests: 9/9 in ~0.2s (A3 container, also green with no NPU devices visible); canary effectiveness verified by mutation (disabling the grammar branch turns it red).
- [x] E2E 1P1D eager + graph FULL_DECODE_ONLY: rows 3-4 above, byte-identical to the lmhead-off control.
- [x] V2 vs V1 performance parity on the same 1P1D shape (table above).
- MTP spec-decode combination: draft-side alignment goes to a follow-up PR, and the combo is currently blocked upstream anyway (#14871); e2e validation will land with that follow-up.

- vLLM version: v0.27.1
- vLLM main: https://github.com/vllm-project/vllm/commit/ba07e4a48fc951300d97eb506217dd530583dea3

### CODE REVIEW

File: vllm_ascend/worker/v2/model_runner.py [hunk 1]
```diff
@@ -489,6 +489,75 @@ def prepare_inputs(
...
+            f"lmhead TP logits rows ({num_logits}) exceed the group-agreed capacity "
+            f"({capacity} = max_num_reqs * decode_query_len); the capacity formula "
+            "no longer matches upstream logits production."
+        )
+
+        sample_hidden_states = hidden_states[input_batch.logits_indices]
+        if num_logits < capacity:
+            sample_hidden_states = torch.nn.functional.pad(sample_hidden_states, (0, 0, 0, capacity - num_logits))
+        logits = self.model.compute_logits(sample_hidden_states)
+        logits = logits[:num_logits]
```
[the remark this answers is not in the document]
Comment (xuchi-0808) [reply vllm_ascend/worker/v2/model_runner.py]: Not taking this suggestion — the premise doesn't hold. The samplers never index `logits` by `logits_indices`; the actual direction is the opposite. In `Sampler.__call__` (vllm/v1/worker/gpu/sample/sampler.py:79-80), `positions` and `input_ids` are gathered **by** `logits_indices` — i.e. the other tensors are compressed down to `num_logits` rows to match the logits. The sampler also explicitly parameterizes on `logits.shape[0]` as the number of logits rows (`expanded_logits = logits.shape[0] != idx_mapping_np.shape[0]`, sampler.py:110); it has no notion of `num_tokens`.

Upstream `sample()` itself produces `(num_logits, vocab)` logits: `sample_hidden_states = hidden_states[input_batch.logits_indices]` → `compute_logits` → `self.sampler(logits, input_batch)` (vllm/v1/worker/gpu/model_runner.py). Verified empirically: with `num_tokens=10` and a 3-entry `logits_indices`, upstream `GPUModelRunner.sample` hands the sampler logits of shape `(3, vocab)`, and this override hands over exactly the same shape — the dispatch-tail canary test in this PR asserts exactly that (bitwise-identical sampler/grammar/rejection call args), and is green.

Reconstructing a full `(num_tokens, vocab)` tensor would actually break the grammar path: `apply_grammar_bitmask` maps requests to rows via `cu_num_logits` offsets (vllm/v1/worker/gpu/structured_outputs.py:43-49), i.e. indices in `[0, total_num_logits)` — logits space, not token space. With a full-batch tensor those offsets would point at the wrong rows, introducing the exact bug this comment tries to prevent.

File: tests/ut/worker/test_model_runner_v2_finegrained_tp.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+    compute_input = runner.model.compute_logits.call_args.args[0]
+    # compute_logits sees the group-agreed capacity, not the real row count
+    assert compute_input.shape == (16, hidden_dim)
+    # real rows are the indexed hidden states, padding rows are zero
+    torch.testing.assert_close(compute_input[:3], hidden_states[indices])
+    assert torch.all(compute_input[3:] == 0)
+    # the sampler only sees the trimmed real rows
+    sampled_logits = runner.sampler.call_args.args[0]
+    assert sampled_logits.shape[0] == 3
```
[the remark this answers is not in the document]
Comment (xuchi-0808) [reply tests/ut/worker/test_model_runner_v2_finegrained_tp.py]: Keeping the assertion as is. The premise is inverted: the sampler never indexes `logits` by `logits_indices` — in `Sampler.__call__` (sampler.py:79-80) it's `positions`/`input_ids` that get gathered by `logits_indices` to match the logits row count. Upstream `sample()` itself hands the sampler `(num_logits, vocab)` logits (it computes them from `hidden_states[logits_indices]`), verified empirically: with `num_tokens=10` and a 3-entry `logits_indices`, upstream passes shape `(3, vocab)`. This assertion locks that same contract for the override, so `3` (the `num_logits` of this test) is the correct expectation.

File: tests/ut/worker/test_model_runner_v2_finegrained_tp.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    input_batch.num_draft_tokens = 5
+    draft_logits = MagicMock()
+    runner.speculator.draft_logits = draft_logits
+
+    with patch("vllm_ascend.worker.v2.model_runner.lmhead_tp_enable", return_value=True):
+        runner.sample(hidden_states, input_batch, None)
+
+    runner.sampler.assert_not_called()
+    args = runner.rejection_sampler.call_args.args
+    assert args[0].shape[0] == 2  # trimmed before rejection sampling
```
[the remark this answers is not in the document]
Comment (xuchi-0808) [reply tests/ut/worker/test_model_runner_v2_finegrained_tp.py]: Same as above — the rejection sampler branch receives the same logits the non-spec branch does, under the same contract: upstream `sample()` produces `(num_logits, vocab)` logits (computed from `hidden_states[logits_indices]`) and passes them straight to `self.rejection_sampler`; nothing downstream expands logits back to token space. The dispatch-tail canary in this PR confirms the override and upstream hand over bitwise-identical arguments on this branch too. `2` is the `num_logits` of this test case, so the assertion is correct.

File: tests/ut/worker/test_model_runner_v2_finegrained_tp.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    grammar_output = MagicMock()
+
+    with patch("vllm_ascend.worker.v2.model_runner.lmhead_tp_enable", return_value=True):
+        runner.sample(hidden_states, input_batch, grammar_output)
+
+    bitmask_logits = runner.structured_outputs_worker.apply_grammar_bitmask.call_args.args[0]
+    sampled_logits = runner.sampler.call_args.args[0]
+    # grammar and sampler operate on the same trimmed logits tensor
+    assert bitmask_logits is sampled_logits
+    assert bitmask_logits.shape[0] == 2
```
[the remark this answers is not in the document]
Comment (xuchi-0808) [reply tests/ut/worker/test_model_runner_v2_finegrained_tp.py]: Keeping `2`. `apply_grammar_bitmask` maps requests to logits rows via `cu_num_logits` offsets (structured_outputs.py:43-49) — indices in `[0, total_num_logits)`, and its scratch buffer is sized `max_num_logits`. The expected shape is therefore `(num_logits, vocab)`, which is what upstream `sample()` produces and what the override preserves after trimming the padding. Asserting the full-batch shape `10` here would lock in the wrong contract (and would misalign the `cu_num_logits`-based offsets the kernel uses).

**Comment by xuchi-0808** [comment]
/rerun
[Bot]: rerun completed.

**Failed:**
- E2E: still in progress, retry /rerun after completion

**Review by maoxx241 — APPROVED** [review]

File: vllm_ascend/worker/v2/model_runner.py [hunk 5]
```diff
@@ -688,6 +688,117 @@ def prepare_inputs(  # type: ignore[misc]
...
+        assert num_logits <= capacity, (
+            f"lmhead TP logits rows ({num_logits}) exceed the group-agreed capacity "
+            f"({capacity} = max_num_reqs * decode_query_len); the capacity formula "
+            "no longer matches upstream logits production."
+        )
+
+        sample_hidden_states = hidden_states[input_batch.logits_indices]
+        if num_logits < capacity:
+            sample_hidden_states = torch.nn.functional.pad(sample_hidden_states, (0, 0, 0, capacity - num_logits))
+        logits = self.model.compute_logits(sample_hidden_states)
```
Comment (weiguihua2) [inline vllm_ascend/worker/v2/model_runner.py:726]: Can super.sample be adjusted?

File: tests/ut/worker/test_model_runner_v2_finegrained_tp.py [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+    with (
+        patch("vllm_ascend.worker.v2.model_runner.lmhead_tp_enable", return_value=True),
+        pytest.raises(AssertionError, match="group-agreed capacity"),
+    ):
+        runner.sample(torch.randn(20, 4), input_batch, None)
+
+    runner.model.compute_logits.assert_not_called()
+
+
+def test_lmhead_tp_sends_trimmed_logits_to_rejection_sampler():
```
Comment (keyi-zz) [inline tests/ut/worker/test_model_runner_v2_finegrained_tp.py]: please check which tests is needed

File: vllm_ascend/worker/v2/model_runner.py (hunk 5 above)
Comment (xuchi-0808) [reply vllm_ascend/worker/v2/model_runner.py:726]: We discussed this offline: upstream sample() bundles compute_logits and the sampler dispatch with no seam between them (no way to pass logits in), and the trim must sit exactly between the two. So the override takes over the whole method, and a dispatch-tail canary UT compares our tail call-by-call against upstream sample() to catch any drift on main bumps.

**Review by keyi-zz — APPROVED** [review]

**Review by weiguihua2 — APPROVED** [review]
